metrics-operator/controllers/common/providers/datadog/datadog.go

```diff
@@ -0,0 +1,82 @@
+package datadog
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	//nolint:gci
+	"net/http" //nolint:gci
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV1"
+	"github.com/go-logr/logr"
+	metricsapi "github.com/keptn/lifecycle-toolkit/metrics-operator/api/v1alpha2"
+	//nolint:gci
+	//nolint:gci
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type KeptnDataDogProvider struct {
+	Log        logr.Logger
+	HttpClient http.Client
+	K8sClient  client.Client
+}
+
+// EvaluateQuery fetches the SLI values from datadog provider
+func (d *KeptnDataDogProvider) EvaluateQuery(ctx context.Context, metric metricsapi.KeptnMetric, provider metricsapi.KeptnMetricsProvider) (string, []byte, error) {
+	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	defer cancel()
+
+	// Assumed default metric duration as 5 minutes
+	// Think a better way to handle this
+	intervalInMin := 5
+	fromTime := time.Now().Add(time.Duration(-intervalInMin) * time.Minute).Unix()
+	toTime := time.Now().Unix()
+	qURL := provider.Spec.TargetServer + "/api/v1/query?from=" + strconv.Itoa(int(fromTime)) + "&to=" + strconv.Itoa(int(toTime)) + "&query=" + url.QueryEscape(metric.Spec.Query)
+	req, err := http.NewRequestWithContext(ctx, "GET", qURL, nil)
+	if err != nil {
+		d.Log.Error(err, "Error while creating request")
+		return "", nil, err
+	}
+
+	apiKeyVal, appKeyVal, err := getDDSecret(ctx, provider, d.K8sClient)
+	if err != nil {
+		return "", nil, err
+	}
+
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Dd-Api-Key", apiKeyVal)
+	req.Header.Set("Dd-Application-Key", appKeyVal)
+
+	res, err := d.HttpClient.Do(req)
+	if err != nil {
+		d.Log.Error(err, "Error while creating request")
+		return "", nil, err
+	}
+	defer func() {
+		err := res.Body.Close()
+		if err != nil {
+			d.Log.Error(err, "Could not close request body")
+		}
+	}()
+
+	b, _ := io.ReadAll(res.Body)
+	result := datadogV1.MetricsQueryResponse{}
+	err = json.Unmarshal(b, &result)
+	if err != nil {
+		d.Log.Error(err, "Error while parsing response")
+		return "", nil, err
+	}
+
+	if len(result.Series) == 0 {
+		d.Log.Info("No values in query result")
+		return "", nil, fmt.Errorf("no values in query result")
+	}
+
+	points := (result.Series)[0].Pointlist
+	value := strconv.FormatFloat(*points[len(points)-1][1], 'g', 5, 64)
```

Comment:
I am not a DD expert, so I might be wrong. This looks like we pick the most recent datapoint the API returns instead of doing some aggregations on the range vector. Looking at #981, we'll introduce a range parameter that we no longer need to hardcode the 5min. With this implementation, we ignore the range vector returned and we only look at the latest value reported. I think we should do some "avg" out of these values, similar to what we do in the DT provider.

Reply:
yes I have thought about it and put this in the PR description. 

> Qoute > I saw in dynatrace implementation, a [getSingleValue](https://github.com/keptn/lifecycle-toolkit/blob/bd493578df8825a52ec0f027583341a80b3c90f6/metrics-operator/controllers/common/providers/dynatrace/dynatrace.go#L82) is implemented. It basically, returns the average of the metric value. Therefore, I think the current logic might need some amendments. Please suggest.

Basically the original Keptn Datadog implementation uses this logic. That's why I picked it up. Ref: [here](https://github.com/keptn-sandbox/datadog-service/blob/10483aa887e7955910d08eab61f3b5e89b5a8632/eventhandlers.go#L139)

Reply:
avg method applied.